### PR TITLE
Allow module-local classes to be loaded by external modules

### LINKED ISSUE #919: Eventual collisions between opaque STL containers
After some chatting on gitter, @jagerman suggested that I raise an issue here, mostly so that it can be kept track of.

This is related to #439, but has more to do with what happens in the long run when more and more packages are developed using pybind11.

Currently,  opaque STL containers are most often generated with a combo of the `PYBIND11_MAKE_OPAQUE` macro and a call to `py::bind_vector`.

If two different modules by different developers generate the same opaque containers, an import error occurs if both modules are used together.  One simple solution is to wrap the binding in try/catch blocks as shown here: https://github.com/molpopgen/opaque_test.  (That example uses ccpimport to compile the code.)

However, the try/catch solution is imperfect: the name seen on the Python side depends on the import order.  Further, there is the deeper problem that `py::bind_vector` may optionally be invoked with `py::buffer_protocol()` and that the class type returned may be further augmented via `.defs` in the usual way.  The possible result then is that both the name and features/behavior depend on import order, which seems less than ideal.

The gitter discussion ended with the agreement that a solution is needed and that it is not clear what that solution should look like.

### PR DESCRIPTION
**Pull request by jagerman**
The main point of `py::module_local` is to make the C++ -> Python cast unique so that returning/casting a C++ instance is well-defined.  Unfortunately it also makes *loading* unique, but this isn't particularly desirable or necessary: when an instance contains `Type` instance there's no reason it shouldn't be possible to pass that instance to a bound function taking a `Type` parameter, even if that function is in another module.

This commit solves the issue by storing the module-local `&type_caster_generic::load` in a global multimap so that we can call into local modules' `load()` methods for `module_local` types.  For module-local types that override a globally-registered types a global type conversion is also attempted.

Fixes #919.

It increases `.so` size slightly (a little over 4KB, not counting the added tests), but as the additions are entirely in non-templated code, that number shouldn't change with the number of bound types.

### CODE REVIEW

**Comment by dean0x7d** [comment]
Just a quick comment for now. The search for an external loader looks really complicated. Instead of the multimap in `internals`, have you considered attaching the external load as a capsule to the Python type object in `generic_type::initialize()`?

**Comment by jagerman** [comment]
That simplifies it; I've pushed a couple of commits to stash things in an attribute of the python type.  The second commit is simpler, but increases the `type_info` size by the size of a member function pointer.

**Comment by dean0x7d** [comment]
I had a go at simplifying this just a bit more: reduced the reliance on caster state (recursion flag, typeinfo state change) because these will need to go anyway with the planned new-style `maybe` loaders. Also switched to a regular function pointer for portability (pointer to member function can be weird). The changes reduce binary size (at least on clang). Everything should still work as before. Let me know if you agree with these changes.

**Comment by jagerman** [comment]
Looks great!

**Comment by jagerman** [comment]
I think we should include this in 2.2 -- I noticed you also put it in the upgrade doc, so I'm guessing you agree.

**Comment by dean0x7d** [comment]
Yes, definitely. It makes `module_local` significantly more convenient (and easier to explain because there's one less caveat). If everything else looks good, feel free to squash and merge this.

**Comment by jagerman** [comment]
The current build is just for the comment change, right?  (i.e. safe to not wait for the ci tests to finish?)

**Comment by dean0x7d** [comment]
Yeah.

File: include/pybind11/cast.h [hunk 1]
```diff
@@ -686,13 +694,32 @@ class type_caster_generic {
...
+    /// native typeinfo, or when the native one wasn't able to produce a value.
+    PYBIND11_NOINLINE bool try_load_foreign_module_local(handle src) {
+        constexpr auto *local_key = "_pybind11_module_local_typeinfo";
+        const auto pytype = src.get_type();
+        if (!hasattr(pytype, local_key))
+            return false;
+
+        type_info *foreign_typeinfo = reinterpret_borrow<capsule>(getattr(pytype, local_key));
+        if (foreign_typeinfo == typeinfo)
+            return false; // Only consider this foreign loader if actually foreign
```
Comment (jagerman) [inline include/pybind11/cast.h]: I think there's a slight problem here: if the type has a masked global registration, `typeinfo` will have been changed to the global one, in which case the `foreign_typeinfo == typeinfo` isn't going to be true, so we could end recursing.

**Comment by jagerman** [comment]
There's another problem here that I'm looking into: We aren't checking that the foreign loader is actually a loader of the right type.

File: include/pybind11/cast.h [hunk 2]
```diff
@@ -655,8 +665,6 @@ class type_caster_generic {
         return inst.release();
     }
 
-protected:
```
Comment (jagerman) [inline include/pybind11/cast.h]: Was this removal intentional?

Comment (dean0x7d) [inline include/pybind11/cast.h]: Yes. Direct access to `value` is needed in the `module_local_load` lambda. And seeing as `detail::type_caster_generic` is already an implementation detail, there isn't really much of a need to protect access.

**Comment by dean0x7d** [comment]
> There's another problem here that I'm looking into: We aren't checking that the foreign loader is actually a loader of the right type.

Good catch.

File: include/pybind11/cast.h (hunk 1 above)
Comment (dean0x7d) [inline include/pybind11/cast.h]: Do you mean a local binding masking a local one? That should work and is covered by the tests. (Not the `cpptype` mismatch mentioned below.)

Comment (jagerman) [inline include/pybind11/cast.h]: I'm thinking:
- the local type masks the global type, is the correct type_info for the given argument, but the load fails anyway.
- the global type gets checked, which also fails (the global type is not right).
- foreign loaders get checked, but because `typeinfo` has been changed to the global one, we don't short-circuit in `try_load_foreign` and so end up recursing.

It's not immediately obvious to me that the first step is impossible; though it's not immediately obvious to me how I can trigger it in a test, either.

Comment (jagerman) [inline include/pybind11/cast.h]: I think 225caf3 ought to solve this: when using a proper function (rather than a lambda) we can just test whether that function is foreign rather than relying on `typeinfo`.

Comment (dean0x7d) [inline include/pybind11/cast.h]: I see, seems reasonable.
